### PR TITLE
fix(merkle-distributor): [H12] track claimed rewards per window

### PR DESCRIPTION
**Pull request by Reinis-FRP**
**Motivation**

Audit identified the following issue:
_The MerkleDistributor unnecessarily allows for separate rewards to be stolen from each other if a single merkle tree allows to withdraw more rewards than were provided to that tree._

**Summary**

This PR fixes identified issue by adding `remainingAmount` parameter to each window that is decreased with each claim. Claims that cause it to underflow are reverting.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 1]
```diff
@@ -248,6 +238,8 @@ contract MerkleDistributor is Ownable {
 
     // Verify claim is valid and mark it as completed in this contract.
     function _verifyAndMarkClaimed(Claim memory _claim) private {
+        // Check claimed rewards do not exceed remaining funding for the window.
+        require(_claim.amount <= merkleWindows[_claim.windowIndex].remainingAmount, "Insufficient funding for window");
```
Comment (chrismaree) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:242]: I dont think this is strictly required as the call on line 175 would revert if there was not enough but it's better to have a more descriptive revert message than just `underflow`

**Review by chrismaree — APPROVED** [review]
this is good, makes sence and is a clean solution. I'm not really sure why we didnt just do this in the first place... if you have a block of 14 lines explaining a known issue in the contracts it should likely be fixed •͡˘㇁•͡˘

@nicholaspai do you see any issues with what we've proposed here? I know you worked closely on this particular part of the code.

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol (hunk 1 above)
Comment (mrice32) [reply packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:242]: Agree. We should drop this as it's an unnecessary gas expenditure.

**Review by mrice32 — APPROVED** [review]
Looks good!

I'm not sure I 100% agree that this is needed.

Before this change, you were trusting the owner to make sure they deposit enough funds such that the sum of all leaves of all roots are funded.

After this change, you're trusting the owner to make sure they deposit enough funds such that the sum of the remainingAmount values are funded.

In the former, one root could steal from another if one had a leaf with a very large amount value. In the latter you could have remainingAmount value that is set very high.

The trust assumptions feel the same IMO. I would argue the only thing that would make this trustless would be to drop the withdrawRewards, deleteWindow methods and require that funds are pulled from the caller in setWindow.

@Reinis-FRP happy to approve as is, but it feels a bit like we're adding gas for no real change in trust assumptions.

**Comment by Reinis-FRP** [comment]
> The trust assumptions feel the same IMO. I would argue the only thing that would make this trustless would be to drop the withdrawRewards, deleteWindow methods and require that funds are pulled from the caller in setWindow.

Agree to your judgement on trust assumptions when owner is EOA, but this considerably improves security when owned by [OptimisticDistributor](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol) as individual sponsors funding rewards might otherwise try to steal each others' deposits in MerkleDistributor by misconfiguring price identifier/ancilary data and making it difficult for honest verifiers to dispute malicious payout proposals (e.g. short OO liveness with high bonds).

Currently OptimisticDistributor does not implement calling `withdrawRewards` and `deleteWindow` methods on its controlled MerkleDistributor, but would be something to consider in future releases (e.g. contract controlled clawback mechanism by sponsors)

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**